### PR TITLE
feat(shell-api): add shardedDataDistribution to sh.status() MONGOSH-1326

### PR DESCRIPTION
**Pull request by gagik**
Considering the balancing of a sharded namespace would be sensitive to orphans and data volume, and in order to count with a mechanism that easily allows checking the current state of the cluster, it would be good if the sh.status() provides the following information per sharded namespaces:

    Estimated number of orphan documents per shard
    Data volume per shard

This is supported in `6.0.3+`

### CODE REVIEW

File: packages/shell-api/src/helpers.ts [hunk 1]
```diff
@@ -644,10 +668,83 @@ export async function getPrintableShardStatus(
...
+          [count: number]:
+            | 'Success'
+            | `Failed with error '${string}', from ${string} to ${string}`;
+        };
+    'Balancer active window is set between'?: `${string} and ${string} server local time`;
+    'Last reported error'?: string;
+    'Time of Reported error'?: string;
+    'Collections with active migrations'?: `${string} started at ${string}`[];
+  };
+  shardedDataDistribution?: ShardedDataDistribution;
```
Comment (gagik) [inline packages/shell-api/src/helpers.ts:733]: I am adding it as a separate field at the moment since that's just a simple solution, but we could make it cleaner/nicer for the user. Open to suggestions

File: packages/shell-api/src/helpers.ts [hunk 2]
```diff
@@ -295,25 +299,27 @@ export async function getPrintableShardStatus(
...
-        ).toArray()) as any[]
+        ).toArray()) as { _id: string; num: number }[]
       ).map((z: { _id: string; num: number }) => {
         return { [z._id]: z.num };
       });
     }
   }
 
-  const balancerRes: Record<string, any> = {};
+  const balancerRes = {} as ShardingStatusResult['balancer'];
```
Comment (lerouxb) [inline packages/shell-api/src/helpers.ts:322]: Why the preference for it this way around as opposed to:

```
const balancerRes: ShardingStatusResult['balancer'] = {};
```

Comment (gagik) [reply packages/shell-api/src/helpers.ts:322]: I'm not a fan of this myself but it'd actually be invalid; we're essentially doing a cast here to an invalid object, kind of breaking type safety; still better than not doing a record. Here's a [playground example of what I mean](https://www.typescriptlang.org/play/?#code/PTAEHUFMBsGMHsC2lQBd5oBYoCoE8AHSAZVgCcBLA1UABWgEM8BzM+AVwDsATAGiwoBnUENANQAd0gAjQRVSQAUCEmYKsTKGYUAbpGF4OY0BoadYKdJMoL+gzAzIoz3UNEiPOofEVKVqAHSKymAAmkYI7NCuqGqcANag8ABmIjQUXrFOKBJMggBcISGgoAC0oACCbvCwDKgU8JkY7p7ehCTkVDQS2E6gnPCxGcwmZqDSTgzxxWWVoASMFmgYkAAeRJTInN3ymj4d-jSCeNsMq-wuoPaOltigAKoASgAywhK7SbGQZIIz5VWCFzSeCrZagNYbChbHaxUDcCjJZLfSDbExIAgUdxkUBIursJzCFJtXydajBVDtOYAXlAAG9FCUStx8ldUJROMwANyKAC+wQQnEENAYADVHKAabSeWJhBVuaLHAFXDSAEQAZhV3IFQrEYrIACYWVVJTytY0dQqyGqjRK6czQOqVabgiocM0PGQvIh4H0GMD2DQvm4zMx2AxmJB+LBoOpEn74HoRF4VQBRVYMRALfQqpLYlXgByoADkwgAcpAJCqgioAPJfMjvQSRrSQI6oRwKVzSPCgJzenTDLD6FAIRDQ4SXIMSH3REQGDjYhZMVgcHhBEpAA)

**Review by lerouxb — APPROVED** [review]
Looks good. I'm no expert in this department, though.

File: packages/shell-api/src/helpers.ts (hunk 2 above)
Comment (lerouxb) [reply packages/shell-api/src/helpers.ts:322]: Wondering if it is better to only assert once it is correct rather than cast. Just a style nit, really.

File: packages/shell-api/src/helpers.ts [hunk 3]
```diff
@@ -644,10 +668,83 @@ export async function getPrintableShardStatus(
...
+    'Balancer active window is set between'?: `${string} and ${string} server local time`;
+    'Last reported error'?: string;
+    'Time of Reported error'?: string;
+    'Collections with active migrations'?: `${string} started at ${string}`[];
+  };
+  shardedDataDistribution?: ShardedDataDistribution;
+  databases: { database: Document; collections: Document }[];
+};
+
+export type ShardedDataDistribution = {
```
Comment (lerouxb) [inline packages/shell-api/src/helpers.ts:737]: Don't know where to leave this comment, but I noticed on the ticket the comment:  "if it's possible I think the order should be ownedSizeBytes, numOwnedDocuments, orphanedSizeBytes, numOrphanedDocs".

Wondering how much effort we should put in for that request. Probably not worth it and kinda weird precedent to be setting. Wouldn't want sorting/ordering to break this when they add fields, for example.

Comment (gagik) [reply packages/shell-api/src/helpers.ts:737]: Yeah, wanted to comment on that too; I think the idea of controlling ordering is weird, I imagine it'd be consistent with whatever order the DB returns and I doubt that'd change much if at all. I will try to run it on an actual and see if that's the case.

File: packages/shell-api/src/helpers.ts (hunk 2 above)
Comment (gagik) [reply packages/shell-api/src/helpers.ts:322]: I think the benefit of the cast is that it gives us IDE hints so at least it adds some checks while the fields are being set; though I agree it also can give misleading idea of what fields are already set depending on context.

Seems fine for initialization though; do you mean like a runtime assertion function?

Comment (lerouxb) [reply packages/shell-api/src/helpers.ts:322]: I just mean that at line 322 it isn't really a ShardingStatusResult['balancer']. But I don't have better ideas.

File: packages/shell-api/src/helpers.ts [hunk 4]
```diff
@@ -644,10 +668,83 @@ export async function getPrintableShardStatus(
...
 
+  let shardedDataDistribution: ShardedDataDistribution | undefined;
+
+  try {
+    // Available since >= 6.0.3
+    const adminDb = db.getSiblingDB('admin');
+    shardedDataDistribution = (await (
+      await adminDb.aggregate([{ $shardedDataDistribution: {} }])
+    ).toArray()) as ShardedDataDistribution;
+  } catch (e) {
```
Comment (addaleax) [inline packages/shell-api/src/helpers.ts:679]:
```suggestion
  } catch {
```

File: packages/shell-api/src/helpers.ts [hunk 5]
```diff
@@ -644,10 +668,83 @@ export async function getPrintableShardStatus(
     )
   ).filter((dbEntry) => !!dbEntry);
 
+  let shardedDataDistribution: ShardedDataDistribution | undefined;
+
+  try {
+    // Available since >= 6.0.3
+    const adminDb = db.getSiblingDB('admin');
+    shardedDataDistribution = (await (
+      await adminDb.aggregate([{ $shardedDataDistribution: {} }])
```
Comment (addaleax) [inline packages/shell-api/src/helpers.ts:677]: Can we do this in parallel to other operations, like we do around ~L527?

File: packages/shell-api/src/helpers.ts [hunk 6]
```diff
@@ -331,13 +337,13 @@ export async function getPrintableShardStatus(
...
       // Is the balancer currently active
-      let balancerRunning = 'unknown';
+      let balancerRunning: 'yes' | 'no' | 'unknown' = 'unknown';
       try {
         const balancerStatus = await configDB.adminCommand({
           balancerStatus: 1,
         });
         balancerRunning = balancerStatus.inBalancerRound ? 'yes' : 'no';
-      } catch (err: any) {
+      } catch (err) {
```
Comment (addaleax) [inline packages/shell-api/src/helpers.ts:346]:
```suggestion
      } catch {
```

File: packages/shell-api/src/helpers.ts [hunk 7]
```diff
@@ -224,10 +224,12 @@ export function processDigestPassword(
...
 export async function getPrintableShardStatus(
-  configDB: Database,
+  db: Database,
   verbose: boolean
-): Promise<Document> {
-  const result = {} as any;
+): Promise<ShardingStatusResult> {
+  const result = {} as ShardingStatusResult;
+
+  const configDB = await getConfigDB(db);
```
Comment (addaleax) [inline packages/shell-api/src/helpers.ts:232]: Uh... this is an unintentional breaking change, no? You can't pass a `configDB` object to `sh.status()` explicitly anymore after this

File: packages/shell-api/src/helpers.ts [hunk 8]
```diff
@@ -364,7 +370,9 @@ export async function getPrintableShardStatus(
       if (activeLocks?.length > 0) {
         balancerRes['Collections with active migrations'] = activeLocks.map(
           (lock) => {
-            return `${lock._id} started at ${lock.when}`;
+            // This type assertion is necessary for the string literal type check
+            // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
+            return `${lock._id} started at ${lock.when}` as `${string} started at ${string}`;
```
Comment (addaleax) [inline packages/shell-api/src/helpers.ts:375]:
```suggestion
            return `${lock._id} started at ${lock.when}` as const;
```

**Review by addaleax — CHANGES_REQUESTED** [review]

File: packages/shell-api/src/helpers.ts (hunk 7 above)
Comment (gagik) [reply packages/shell-api/src/helpers.ts:232]: Oh right, my bad, somehow I thought the `getConfigDB` was always happening already.

I suppose if `configDB` is passed explicitly then we won't end up doing shardedDataDistribution? or does `getSiblingDb` imply one can get `admin` db from a given `config` db?

File: packages/shell-api/src/change-stream-cursor.spec.ts [hunk 9]
```diff
@@ -342,7 +342,6 @@ describe('ChangeStreamCursor', function () {
     it('isExhausted fails', function () {
       try {
         cursor.isExhausted();
-        expect.fail('missed exception');
```
Comment (addaleax) [inline packages/shell-api/src/change-stream-cursor.spec.ts:345]: Unrelated change slipped in here?

File: packages/shell-api/src/helpers.ts [hunk 10]
```diff
@@ -472,11 +491,12 @@ export async function getPrintableShardStatus(
...
             ])
-          ).toArray()
+          ).toArray()) as MigrationResult[]
         );
 
-        const migrationsRes: Record<number, string> = {};
-        migrations.forEach((x: any) => {
+        const migrationsRes: ShardingStatusResult['balancer']['Migration Results for the last 24 hours'] =
+          {};
+        migrations.forEach((x) => {
```
Comment (addaleax) [inline packages/shell-api/src/helpers.ts:499]: Just a note for the future, but if you run into existing `.forEach()` into the future, don't be shy to refactor them to `for ... of` 🙂

**Review by addaleax — APPROVED** [review]
